### PR TITLE
Added batch update for statuses

### LINKED ISSUE #347: Support for Batch Operations
Support for batches, e.g.:
- Batch issuance of credentials
- Batch status changes (revocation)
- Batch credential refresh
- ... more!

### PR DESCRIPTION
**Pull request by andresuribe87**
# Overview
This PR adds a batch endpoint for updating statuses of credentials. It is the last PR that fixes #347

# Description
Added the `v1/credentials/status/batch` endpoint. See the swagger spec file for documentation on the request/response schemas.

# How Has This Been Tested?
Unit and integration tests.

### CODE REVIEW

File: pkg/server/router/credential.go [hunk 1]
```diff
@@ -339,6 +339,72 @@ type UpdateCredentialStatusResponse struct {
 	Suspended bool `json:"suspended"`
 }
 
+type SingleUpdateCredentialStatusRequest struct {
```
Comment (decentralgabe) [inline pkg/server/router/credential.go:342]:
```suggestion
type UpdateCredentialStatusRequest struct {
```

File: pkg/server/router/credential.go [hunk 2]
```diff
@@ -339,6 +339,72 @@ type UpdateCredentialStatusResponse struct {
...
+	// ID of the credential who's status should be updated.
+	ID string `json:"id"`
```
Comment (decentralgabe) [inline pkg/server/router/credential.go:344]: this should be required

File: pkg/server/router/credential.go [hunk 3]
```diff
@@ -339,6 +339,72 @@ type UpdateCredentialStatusResponse struct {
...
+	Requests []SingleUpdateCredentialStatusRequest `json:"requests" maxItems:"100" validate:"required,dive"`
+}
+
+func (r BatchUpdateCredentialStatusRequest) toServiceRequest() *credential.BatchUpdateCredentialStatusRequest {
+	var req credential.BatchUpdateCredentialStatusRequest
+	for _, routerReq := range r.Requests {
+		serviceReq := routerReq.toServiceRequest(routerReq.ID)
+		req.Requests = append(req.Requests, serviceReq)
+	}
+	return &req
```
Comment (decentralgabe) [inline pkg/server/router/credential.go:359]: why return a ptr?

File: pkg/server/server.go [hunk 4]
```diff
@@ -218,6 +218,7 @@ func CredentialAPI(rg *gin.RouterGroup, service svcframework.Service, webhookSer
 	// Credential Status
 	credentialAPI.GET("/:id"+StatusPrefix, credRouter.GetCredentialStatus)
 	credentialAPI.PUT("/:id"+StatusPrefix, credRouter.UpdateCredentialStatus)
+	credentialAPI.PUT(StatusPrefix+"/batch", credRouter.BatchUpdateCredentialStatus)
```
Comment (decentralgabe) [inline pkg/server/server.go:221]: change `/batch` to a const which can be used elsewhere

File: pkg/server/server_credential_test.go [hunk 5]
```diff
@@ -31,6 +31,109 @@ func TestCredentialAPI(t *testing.T) {
 
 	for _, test := range testutil.TestDatabases {
 		t.Run(test.Name, func(tt *testing.T) {
+			tt.Run("Batch Update Credential Status", func(ttt *testing.T) {
```
Comment (decentralgabe) [inline pkg/server/server_credential_test.go:34]: we need a few different test cases here

1. batch of 0
2. batch of 1-n (all valid)
3. batch of 1-n (mixed invalid and valid credentials, invalid = cred doesn't exist, cred status already set)

Another test we might need here and elsewhere is performing an update where the key no longer exists (revoked/deleted)

File: pkg/service/credential/model.go [hunk 6]
```diff
@@ -85,8 +85,22 @@ type UpdateCredentialStatusRequest struct {
...
 
 type UpdateCredentialStatusResponse struct {
-	Revoked   bool `json:"revoked" validate:"required"`
-	Suspended bool `json:"suspended" validate:"required"`
+	Status
+}
+
+type Status struct {
+	// ID of the credentials who's status this object represents.
+	ID        string `json:"id,omitempty"`
```
Comment (decentralgabe) [inline pkg/service/credential/model.go:93]: not required?

File: pkg/service/credential/service.go [hunk 7]
```diff
@@ -675,3 +658,71 @@ func (s Service) BatchCreateCredentials(ctx context.Context, batchRequest BatchC
...
+	return batchResponse, nil
+}
+
+func (s Service) statusListCredentialWatchKey(ctx context.Context, id string) (*storage.WatchKey, error) {
+	gotCred, err := s.storage.GetCredential(ctx, id)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading credential")
+	}
+
+	if gotCred.Credential.CredentialStatus == nil {
```
Comment (decentralgabe) [inline pkg/service/credential/service.go:708]: I would recommend a new helper like `HasCredentialStatus` which can also check that `gotCred` is not nil and `gotCred.Credential` is not nil, or else we're at risk for NPEs

File: pkg/service/credential/service.go [hunk 8]
```diff
@@ -675,3 +658,71 @@ func (s Service) BatchCreateCredentials(ctx context.Context, batchRequest BatchC
...
+		return nil, sdkutil.LoggingNewErrorf("credential %q has no credentialStatus field", gotCred.LocalCredentialID)
+	}
+
+	statusPurpose := gotCred.Credential.CredentialStatus.(map[string]any)["statusPurpose"].(string)
```
Comment (decentralgabe) [inline pkg/service/credential/service.go:712]: `GetStatusPurpose` method would be helpful

**Review by decentralgabe — CHANGES_REQUESTED** [review]
see comments

File: pkg/server/server.go (hunk 4 above)
Comment (andresuribe87) [reply pkg/server/server.go:221]: Done

File: pkg/server/server_credential_test.go (hunk 5 above)
Comment (andresuribe87) [reply pkg/server/server_credential_test.go:34]: Added tests for all batch behaviors.

> Another test we might need here and elsewhere is performing an update where the key no longer exists (revoked/deleted)

I'm unsure about some of the details of the behavior you're trying to test. Mind creating an issue and we can discuss?

File: pkg/service/credential/model.go (hunk 6 above)
Comment (andresuribe87) [reply pkg/service/credential/model.go:93]: No, the usage in `UpdateCredentialStatusResponse` is optional.

**Review by decentralgabe — APPROVED** [review]

**Pull request merged**